articles/nginx-reverse-proxy/index.md

```diff
@@ -0,0 +1,161 @@
+**NOTE:** Do not run your application on Port 80 or 443. It will be explained later why this must not be done.  
+Refer the official ExpressJS documentation for Getting Started. [Docs](https://expressjs.com/en/starter/installing.html)
+2. I have installed NGINX on my local machine, but the same could be done on any Virtual Machine where the applications are expected to be deployed.  
+Here is the documentation on how to install NGINX on your machine. [Docs](https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-open-source/) . 
+
+## Step 1: Start two apps running in different ports
+As I've mentioned earlier, I've got two Node.js Apps running on two different ports as shown below.  
+
+**Server app running on Port 3000**
+
+![image](./server1.png)  
+
+**Client app running on Port 3001**  
+
+![image](./client1.png)
+
+Now that we have our apps up and running, we don't want our users to use these applications by typing their PORTS explicitly, so we need to map it with something which is more human-readable. In this example, I will be using subdomains to distinguish between them. Again one is free to use whichever distinguishing element suitable as per their requirement.
+Another example could be a particular route like domain/client and domain/server. The only condition for the distinguishing element is to follow a Valid URL Regular Expression. Learn about Regex from [here](https://regexr.com/).
+
+
+## Step 2: Add DNS records
+This is the part where one would add the DNS records in their DNS management dashboard. If you are running Nginx locally, you can skip this step.
+
+The general DNS Configurations would be something like
+- Server app mapped to the server.domain
+- Client app mapped to the client.domain
+
+My Localhost Config, in this case, would be
+- Server mapped to server.localhost
+- Client mapped to client.localhost
+
+There are two standard protocols HTTP and HTTPS. The default port for HTTP is 80 and HTTPS is 443. This is the reason we must not run our applications on these ports because our NGINX server would be running on these two ports. All the requests the client makes would either be redirected to Port 80 or 443 from where it would be redirected internally to the corresponding application.
+
+## Step 4 - Configure NGINX at 80 for HTTP and 443 for HTTPS
+Now that we have our apps running and our DNS records ready. We can start configuring our NGINX Reverse Proxy to make it all work.
+Navigate into the Nginx folder and locate  ```nginx.conf``` .
+
+Add these configurations inside the HTTP block.
+
+### Step 4.1 - HTTP
+
+```
+server {       
+     listen       80;
+     server_name  server.domain;
+     location / {
+          proxy_pass "http://localhost:3000" ;
+     }
+}
+server {       
+     listen       80;
+     server_name  client.domain;
+
+     location / {
+          proxy_pass "http://localhost:3001" ;
+     }
+}
+```
+
+### Step 4.2 - HTTPS
+
+```
+server{
```

Comment:
Adding this snippet resulted in a error when testing the configuration `nginx -t`. `nginx: [emerg] unexpected "}"`. Shouldn't there be a space between server and the first { as in the HTTP code snippet above?

Reply:
It doesn't throw any error for me, if adding a space fixes the problem. I'll edit the code snippet in the article.

Reply:
It doesn't throw any error for me, if adding a space fixes the problem. I'll edit the code snippet in the article.